Packages/com.unity.inputsystem/InputSystem/InputManager.cs

```diff
@@ -2952,371 +2952,393 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
-                        DelegateHelpers.InvokeCallbacksSafe(ref m_DeviceChangeListeners,
-                            device, InputDeviceChange.ConfigurationChanged, "InputSystem.onDeviceChange");
-                        break;
+                        case DeviceConfigurationEvent.Type:
+                            device.NotifyConfigurationChanged();
+                            InputActionState.OnDeviceChange(device, InputDeviceChange.ConfigurationChanged);
+                            DelegateHelpers.InvokeCallbacksSafe(ref m_DeviceChangeListeners,
+                                device, InputDeviceChange.ConfigurationChanged, "InputSystem.onDeviceChange");
+                            break;
 
-                    case DeviceResetEvent.Type:
-                        ResetDevice(device, alsoResetDontResetControls: ((DeviceResetEvent*)currentEventReadPtr)->hardReset);
-                        break;
-                }
+                        case DeviceResetEvent.Type:
+                            ResetDevice(device,
+                                alsoResetDontResetControls: ((DeviceResetEvent*)currentEventReadPtr)->hardReset);
+                            break;
+                    }
 
-                m_InputEventStream.Advance(leaveEventInBuffer: leaveInBuffer);
-            }
+                    m_InputEventStream.Advance(leaveEventInBuffer: leaveInBuffer);
+                }
 
-            m_Metrics.totalEventProcessingTime += ((double)(Stopwatch.GetTimestamp() - processingStartTime)) / Stopwatch.Frequency;
-            m_Metrics.totalEventLagTime += totalEventLag;
+                m_Metrics.totalEventProcessingTime +=
+                    ((double)(Stopwatch.GetTimestamp() - processingStartTime)) / Stopwatch.Frequency;
+                m_Metrics.totalEventLagTime += totalEventLag;
 
-            // Remember how much data we retained so that we don't count it against the next
-            // batch of events that we receive.
-            InputUpdate.s_LastUpdateRetainedEventCount = (uint)m_InputEventStream.numEventsRetainedInBuffer;
-            InputUpdate.s_LastUpdateRetainedEventBytes = m_InputEventStream.numBytesRetainedInBuffer;
+                // Remember how much data we retained so that we don't count it against the next
+                // batch of events that we receive.
+                InputUpdate.s_LastUpdateRetainedEventCount = (uint)m_InputEventStream.numEventsRetainedInBuffer;
+                InputUpdate.s_LastUpdateRetainedEventBytes = m_InputEventStream.numBytesRetainedInBuffer;
 
-            m_InputEventStream.Close(ref eventBuffer);
+                m_InputEventStream.Close(ref eventBuffer);
+            }
+            catch (Exception)
+            {
+                // We need to restore m_InputEventStream to a sound state
+                // to avoid failing recursive OnUpdate check next frame.
+                m_InputEventStream.CleanUpAfterException();
+                throw;
+            }
```

Comment:
Hmm, I'm wondering now if it makes more sense to replace this catch with a finally block and move the input event stream close method into it, delete the CleanUpAfterException method, and let the NativeInputRuntime do its job and reset the event buffer if there's an exception.

Reply:
Mmm, maybe, though data correctness after exception is not a given, so I would probably would still put a more exception friendly code branch in Close, e.g. don't reuse existing buffers and just reset ..

Though it's a bit of a slippery slope, our best effort here should be to recover, but we're likely can get in situations were the only recover procedure is to restart whole editor/player ...